nntrainer/include/delegate.h

```diff
@@ -0,0 +1,179 @@
+   *
+   * @param num_threads Number of threads
+   */
+  void setNumThreads(unsigned int num_threads);
+
+private:
+  /**
+   * @brief backend type stored
+   */
+  BackendType backend;
+
+  /**
+   * @brief number of threads stored
+   */
+  unsigned int num_threads;
+};
+
+/**
+ * @class Device
+ * @brief Device to be used for the operations to run.
+ * @note The selected delegate device will restrict the supported backend.
+ *
+ */
+class Device {
+public:
+  /**
+   * @brief Device type enumeration
+   */
+  enum class DeviceType {
+    cpu, /**< CPU as the device */
+    gpu, /**< GPU as the device */
+    npu  /**< NPU as the device */
+  };
+
+  /**
+   * @brief Construct a new Device object
+   *
+   * @param device Device delegate to be used for the operations to run. Default
+   * device is cpu.
+   */
+  Device(DeviceType device = DeviceType::cpu) :
+    device(device),
+    memory_fraction(1.0),
+    soft_placement(false),
+    precision_loss(false) {}
+
+  /**
+   * @brief Set the Device object
+   *
+   * @param device The device to be set to
+   */
+  void setDevice(DeviceType device) { this->device = device; }
+
+  /**
+   * @brief Get the Device object
+   *
+   * @return DeviceType The device type which is set
+   */
+  DeviceType getDevice() { return device; }
+
+  /**
+   * @brief Set the maximum memory fraction which can be used by the framework
+   *
+   * @param memory_fraction Maximum fraction of memory to be used
+   */
+  void setMemoryFraction(float memory_fraction) {
+    throw std::logic_error("Not yet implemented");
+  }
+
+  /**
+   * @brief Allow placing the operations on device softly
+   * @details This allows framework to use some other device other than the set
+   * device for some of the operations for optimization or if an operation is
+   * not supported for the selected device. This might incur extra memory copy
+   * overhead.
+   * @note When set to false, constructing the model can result in error if a
+   * required operation is not avaiable for that device.
+   *
+   * @param soft_placement True to allow soft placement, else false
+   */
+  void allowSoftPlacement(bool soft_placement) {
+    throw std::logic_error("Not yet implemented");
+    this->soft_placement = soft_placement;
+  }
+
+  /**
+   * @brief Allow using low precision version of some of the operations
+   * @details This allows framework to use low precision operations (like
+   * float16 instead of float32) for some of the operations for higher
+   * performance with minimum impact to performance
+   *
+   * @param precision_loss True to allow loss in precision, else false
+   */
+  void allowPrecisionLoss(bool precision_loss) {
+    throw std::logic_error("Not yet implemented");
+    this->precision_loss = precision_loss;
+  }
+
+private:
+  DeviceType device;     /**< Device type for the object */
+  float memory_fraction; /**< Maximum memory fraction which can be used by the
```

Comment:
Can this be double?

Reply:
This can be double.
However, this is still experimental and I don't expect users to specify more than 2 places decimal value (precision of lower than 0.01 will be rare).